internal/tools/protoc-gen-consul-rate-limit/main.go

```diff
@@ -0,0 +1,139 @@
+// protoc-gen-consul-rate-limit maintains the mapping of gRPC method names to
+// a specification of how they should be rate-limited. This is used by the gRPC
+// InTapHandle function (see agent/grpc-middleware/rate.go) to enforce relevant
+// limits without having to call the handler.
+//
+// It works in two phases:
+//
+//	1. Buf/protoc invokes this plugin for each .proto file. We extract the rate
+//	   limit specification from an annotation on the RPC:
+//
+//	   service Foo {
+//	     rpc Bar(BarRequest) returns (BarResponse) {
+//	       option (hashicorp.consul.internal.ratelimit.spec) = {
+//	         operation_type: OPERATION_TYPE_WRITE,
+//	       };
+//	     }
+//	   }
+//
+//	   We write a JSON array of the limits to protobuf/package/path/.ratelimit.tmp:
+//
+//	   [
+//	     {
+//	       "MethodName": "/Foo/Bar",
+//	       "OperationType": "OPERATION_TYPE_WRITE",
+//	     }
+//	   ]
+//
+//	2. The protobuf.sh script (invoked by make proto) runs our postprocess script
+//	   which reads all of the .ratelimit.tmp files in proto and proto-public and
+//	   generates a single Go map in agent/grpc-middleware/rate_limit_mappings.gen.go
+package main
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"google.golang.org/protobuf/compiler/protogen"
+	"google.golang.org/protobuf/proto"
+
+	"github.com/hashicorp/consul/proto-public/annotations/ratelimit"
+)
+
+const (
+	outputFileName = ".ratelimit.tmp"
+
```

Comment:
I'm curious why `.ratelimit.tmp` is a hidden file and is it intentional that they never get deleted? I'd usually tie this sort of thing back to a `make clean` but that doesn't exist :laugh: